services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/handlers/VeniceServerGrpcHandler.java

```diff
@@ -0,0 +1,41 @@
+package com.linkedin.venice.listener.grpc.handlers;
+
+import com.linkedin.venice.listener.grpc.GrpcHandlerContext;
+import com.linkedin.venice.protocols.VeniceServerResponse;
+import io.grpc.stub.StreamObserver;
+
+
+public abstract class VeniceServerGrpcHandler {
+  protected VeniceServerGrpcHandler next;
+
+  public VeniceServerGrpcHandler() {
+  }
+
+  public void addNextHandler(VeniceServerGrpcHandler next) {
+    VeniceServerGrpcHandler current = this;
+    while (current.next != null) {
+      current = current.next;
+    }
+
+    current.next = next;
+  }
+
+  public abstract void processNext(GrpcHandlerContext ctx);
+
+  public void sendError(GrpcHandlerContext ctx) {
```

Comment:
I think we can get rid of sendError since in each processNext we are checking ctx.hasError

Reply:
i think we should not do the check. We can pass the request along via sendError and send the request back in the last handler (outbound stats). This way we can still achieve the same functionality where we record metrics for failed/errored requests.

Reply:
Yeah let's keep do one of the these: check ctx has error set or not or just use next.sendError